src/Lean/Meta/Constructions.lean

```diff
@@ -7,32 +7,100 @@ prelude
-  let env ← ofExceptKernelException (f (← getEnv) declName)
-  modifyEnv fun _ => env
+def mkRecOn (declName : Name) : MetaM Unit := do
+  let name := mkRecOnName declName
+  let decl ← ofExceptKernelException (mkRecOnImp (← getEnv) declName)
+  addDecl decl
+  setReducibleAttribute name
+  modifyEnv fun env => markAuxRecursor env name
+  modifyEnv fun env => addProtected env name
 
-def mkCasesOn (declName : Name) : m Unit := adaptFn mkCasesOnImp declName
-def mkRecOn (declName : Name) : m Unit := adaptFn mkRecOnImp declName
-def mkNoConfusionCore (declName : Name) : m Unit := adaptFn mkNoConfusionCoreImp declName
-def mkBelow (declName : Name) : m Unit := adaptFn mkBelowImp declName
-def mkIBelow (declName : Name) : m Unit := adaptFn mkIBelowImp declName
-def mkBRecOn (declName : Name) : m Unit := adaptFn mkBRecOnImp declName
-def mkBInductionOn (declName : Name) : m Unit := adaptFn mkBInductionOnImp declName
+def mkCasesOn (declName : Name) : MetaM Unit := do
+  let name := mkCasesOnName declName
+  let decl ← ofExceptKernelException (mkCasesOnImp (← getEnv) declName)
+  addDecl decl
+  setReducibleAttribute name
+  modifyEnv fun env => markAuxRecursor env name
+  modifyEnv fun env => addProtected env name
 
-open Meta
+private def mkBelowOrIBelow (declName : Name) (ibelow : Bool) : MetaM Unit := do
+  let .inductInfo indVal ← getConstInfo declName | return
+  unless indVal.isRec do return
+  if ← isPropFormerType indVal.type then return
+
+  let decl ← ofExceptKernelException (mkBelowImp (← getEnv) declName ibelow)
+  let name := decl.definitionVal!.name
+  addDecl decl
+  setReducibleAttribute name
+  modifyEnv fun env => addToCompletionBlackList env name
+  modifyEnv fun env => addProtected env name
+
+def mkBelow (declName : Name) : MetaM Unit := mkBelowOrIBelow declName true
+def mkIBelow (declName : Name) : MetaM Unit := mkBelowOrIBelow declName false
+
+private def mkBRecOrBInductionOn (declName : Name) (ind : Bool) : MetaM Unit := do
+  let .inductInfo indVal ← getConstInfo declName | return
+  unless indVal.isRec do return
+  if ← isPropFormerType indVal.type then return
+  let .recInfo recInfo ← getConstInfo (mkRecName declName) | return
+  unless recInfo.numMotives = indVal.all.length do
+    /-
+    The mutual declaration containing `n` contains nested inductive datatypes.
+    We don't support this kind of declaration here yet. We will probably never will :)
+    To support it, we will need to generate an auxiliary `below` for each nested inductive
+    type since their default `below` is not good here. For example, at
+    ```
+    inductive term
+    | var : string -> term
+    | app : string -> list term -> term
+    ```
+    The `list.below` is not useful since it will not allow us to recurse over the nested terms.
+    We need to generate another one using the auxiliary recursor `term.rec_1` for `list term`.
```

Comment:
Lean's casing rules have changed since this was written, the comment might as well reflect that.
```suggestion
    inductive Term
    | var : String -> Term
    | app : String -> List Term -> Term
    ```
    The `List.below` is not useful since it will not allow us to recurse over the nested terms.
    We need to generate another one using the auxiliary recursor `Term.rec_1` for `List Term`.
```